demos/groktunnel/main.go

```diff
@@ -0,0 +1,125 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"strings"
+	"sync"
+	"time"
+
+	vhost "github.com/inconshreveable/go-vhost"
+	shortuuid "github.com/lithammer/shortuuid/v3"
+	"github.com/progrium/qmux/golang/session"
+)
+
+func fatal(err error) {
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func main() {
+	var port = flag.String("p", "9999", "server port to use")
+	var host = flag.String("h", "vcap.me", "server hostname to use")
+
+	flag.Parse()
+
+	if flag.Arg(0) != "" {
+		conn, err := net.Dial("tcp", "_new."+net.JoinHostPort(*host, *port))
+		fatal(err)
+		client := httputil.NewClientConn(conn, bufio.NewReader(conn))
+		req, err := http.NewRequest("GET", "/", nil)
+		req.Host = "_new." + net.JoinHostPort(*host, *port)
+		fatal(err)
+		client.Write(req)
+		resp, _ := client.Read(req)
+		fmt.Printf("port %s available at:\nhttp://%s\n", flag.Arg(0), resp.Header.Get("X-Public-Host"))
+		c, _ := client.Hijack()
+		sess := session.New(c)
+		for {
+			ch, err := sess.Accept()
+			fatal(err)
+			conn, err := net.Dial("tcp", "127.0.0.1:"+flag.Arg(0))
+			fatal(err)
+			go join(conn, ch)
+		}
+		return
+	}
+
+	l, err := net.Listen("tcp", net.JoinHostPort("0.0.0.0", *port))
```

Comment:
Should this be limited to localhost for the demo? I guess if you were to productionize this, another flag to set the host here would be helpful, but listening on all IPs by default could expose people to unwanted connections.
```suggestion
	l, err := net.Listen("tcp", net.JoinHostPort("127.0.0.1", *port))
```

Reply:
i did want this to work in cases where deployed to digitalocean or something. will add bind addr flag